src/bsoncxx/string/view_or_value.cpp

```diff
@@ -0,0 +1,52 @@
+// Copyright 2015 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <bsoncxx/config/prelude.hpp>
+
+#include <bsoncxx/string/view_or_value.hpp>
+
+namespace bsoncxx {
+BSONCXX_INLINE_NAMESPACE_BEGIN
+namespace string {
+
+const char* view_or_value::c_str() {
+    // If we do not own our string, we cannot guarantee that it is null-terminated,
+    // so make an owned copy.
+    if (!_value) {
+        _value = std::move(_view.to_string());
+        _view = *_value;
+    }
+
+    return _value->c_str();
+}
+
+bool operator==(view_or_value lhs, const char* rhs) {
```

Comment:
I think all of these should be inline. In particular, this one, since if the `const char *` is a literal, we want the compiler to be able to see that we are making a `string_view` of it so it can (hopefully) elide the call to `strlen`.